keps/sig-testing/4224-replace-kubeup-clusters-with-kops/README.md

```diff
@@ -0,0 +1,429 @@
+
+<!--
+
+Production readiness reviews are intended to ensure that features merging into
+Kubernetes are observable, scalable and supportable; can be safely operated in
+production environments, and can be disabled or rolled back in the event they
+cause increased failures in production. See more in the PRR KEP at
+https://git.k8s.io/enhancements/keps/sig-architecture/1194-prod-readiness.
+
+The production readiness review questionnaire must be completed and approved
+for the KEP to move to `implementable` status and be included in the release.
+
+In some cases, the questions below should also have answers in `kep.yaml`. This
+is to enable automation to verify the presence of the review, and to reduce review
+burden and latency.
+
+The KEP must have a approver from the
+[`prod-readiness-approvers`](http://git.k8s.io/enhancements/OWNERS_ALIASES)
+team. Please reach out on the
+[#prod-readiness](https://kubernetes.slack.com/archives/CPNHUMN74) channel if
+you need any help or guidance.
+-->
+
+Not applicable
+
+### Feature Enablement and Rollback
+
+Not Applicable
+
+
+### Dependencies
+
+<!--
+This section must be completed when targeting beta to a release.
+-->
+
+This KEP depends on the following changes in:
+- k/cloud-provider-gcp
+  - images published for both architectures and promoted properly to registry.k8s.io
+  - auth-provider-gcp to be published properly and available at artifacts.k8s.io
+- k/kops
+  - squash various bugs with kops
+
+## Scalability
+
+kops is already used to run scale tests on AWS. We can use it to replace the kube-up scale tests.
+
+### Troubleshooting
+
+<!--
+This section must be completed when targeting beta to a release.
+
+For GA, this section is required: approvers should be able to confirm the
+previous answers based on experience in the field.
+
+The Troubleshooting section currently serves the `Playbook` role. We may consider
+splitting it into a dedicated `Playbook` document (potentially with some monitoring
+details). For now, we leave it here.
+-->
+
+Not Applicable
+
+## Implementation History
+
+<!--
+Major milestones in the lifecycle of a KEP should be tracked in this section.
+Major milestones might include:
+- the `Summary` and `Motivation` sections being merged, signaling SIG acceptance
+- the `Proposal` section being merged, signaling agreement on a proposed design
+- the date implementation started
+- the first Kubernetes release where an initial version of the KEP was available
+- the version of Kubernetes where the KEP graduated to general availability
+- when the KEP was retired or superseded
+-->
+
+## Drawbacks
+
+<!--
+Why should this KEP _not_ be implemented?
+-->
+
+No reason really.
+
+## Alternatives
+
+<!--
+What other approaches did you consider, and why did you rule them out? These do
+not need to be as detailed as the proposal, but should include enough
+information to express the idea and why it was not acceptable.
+-->
+
+We don't have one.
```

Comment:
Was using Cluster API considered? I remember there was an issue open in k/k (can't seem to find the link about deprecating the cluster/ directory and using cluster-api for tests

kOps doesn't support as many infrastructure providers (everything besides AWS and GCE are alpha or beta) so it may be less versatile for running k/k tests across different clouds in the future.

Reply:
cc @justinsb @fabriziopandini @vincepri

Reply:
I didn't articulate this, but CAPI requires a kubernetes cluster to create the test cluster which adds an extra complication.

Also, CAPG isn't well maintained. This is what I found when I took CAPG for a spin at the end of 2022 https://github.com/vmware-tanzu/crash-diagnostics/pull/243

Reply:
Relevant historical slack thread:

https://kubernetes.slack.com/archives/C2C40FMNF/p1657180208598469

Google docs link to previous discussion:

https://docs.google.com/document/d/1n1Znf-SY85Bjo_B_ReKsqAvEbU4Z8Vw-AvgeiSGVXBA/edit#heading=h.xk6zt7zfht6s

Looks like this never got an owner, lemme know if I can help with that.

Reply:
> I didn't articulate this, but CAPI requires a kubernetes cluster to create the test cluster which adds an extra complication

A common way to get around this is to use a kind cluster as bootstrap cluster. All the k/k cloud-provider Azure tests create test k8s clusters using CAPZ currently: https://testgrid.k8s.io/provider-azure-master-signal

 > Also, CAPG isn't well maintained

@dims @richardcase @cpanato are the maintainers of CAPG, they can comment on the project maturity. 

An alternative would be to run most tests using Docker which doesn't require spinning up any cloud infrastructure (less $$ and faster), this how core Cluster API runs all its tests https://github.com/kubernetes-sigs/cluster-api/blob/main/test/infrastructure/docker/README.md

Reply:
i've not been able to spend time on CAPZ ... not that we have a bunch of jobs using `kubetest2-ec2` i feel less inclined to spend time on CAPZ. Also having spent time on `kops` with scale testing, i feel better using `kops` especially since it works well for both EC2 and GCE

Reply:
Setting aside CAPZ (Azure), it doesn't seem like the stated requirements  for a kube-up replacement include any opinions about "cloud provider neutral", which Cluster API is (IME) uniquely setup to fulfill w/ its abstract interface architecture and pluggable cloud provider implementations.

Is that interpretation correct? Should we consider "cloud provider neutral" as a consideration, or is that out-of-scope for the problem at hand?

Reply:
I think that we should consider CAPI, as an option not mutually exclusive to kOps

We have providers for all the cloud providers, we have a docker provider, we have a kubemark provider for scale tests.

Bootstrap is not an issue, you can use kind or even lighter options like https://github.com/fabriziopandini/kBB-8 (8s to start provisioning a cluster)

I think the CAPI community will be available to do its part, for the providers we can discuss about staffing, but at least IMO this is a discussion that we must have

Reply:
> I think the CAPI community will be available to do its part, for the providers we can discuss about staffing, but at least IMO this is a discussion that we must have

+1 Honestly, I haven't explored CAPI properly yet.

> Setting aside CAPZ (Azure), it doesn't seem like the stated requirements for a kube-up replacement include any opinions about "cloud provider neutral", which Cluster API is (IME) uniquely setup to fulfill w/ its abstract interface architecture and pluggable cloud provider implementations.

kube-up provisions clusters for GCP only. If I want to get rid of kube-up, I need to pick something else that provisions clusters on GCP. This can be kops or CAPG

Right now, I'm busy squashing cloud-provider-gcp test failures/bugs(missing kubelet cred provider, bugs with creating Loadbalancers, kube-addon tests)


> Is that interpretation correct? Should we consider "cloud provider neutral" as a consideration, or is that out-of-scope for the problem at hand?

Out of scope for this issue. We already use KinD to run kubernetes e2e tests that don't rely on the cloud and functionality that can be tested with DinD

Reply:
I think this thread is missing the point.

**This KEP is about eliminating kube-up and kube-up jobs, by shifting to existing production-grade coverage.**

Other CI coverage will continue to exist alongside this and folks are welcome to continue to invest in that.

kOps has *already* for year provided reliable CI coverage for the vendors that actually provide us credits.

Future cloud vendors is not the problem, we're many many years into the credits program and moving CI out of google.com, and yet GCP+AWS are the only vendors that actually provide credits. Of those two, with apologies, CAPG support is just not there.


> Bootstrap is not an issue, you can use kind or even lighter options like https://github.com/fabriziopandini/kBB-8 (8s to start provisioning a cluster)

As a kind maintainer ... that's actually unacceptably expensive. KIND is only cheaper when we don't have to run the cloud cluster. If we have to run a kind cluster for every cloud e2e cluster the CI costs are going to get out of hand.

e2e tests can run for a long time and generally require almost no resources in the CI cluster, only the cluster under tests.

> An alternative would be to run most tests using Docker which doesn't require spinning up any cloud infrastructure (less $$ and faster), this how core Cluster API runs all its tests https://github.com/kubernetes-sigs/cluster-api/blob/main/test/infrastructure/docker/README.md

Again, actually much more expensive than kOps. The kind clusters are not free either, just cheaper than a cloud cluster. But one cloud cluster is cheaper than two cloud clusters or a cloud cluster + a kind cluster.

Reply:
Quick note, lately kOps has gotten some reasonable support for Azure, though did not get a chance to check if there is any interest in running automated tests there, or any Azure credits.

Also, @justinsb had a thought about adding CAPI support in the future. See https://github.com/kubernetes/kops/pull/15522.

Reply:
Thanks for raising this point folks, I generally agree with most comments here, and at large I'd echo what @BenTheElder mentioned above.

While CAPI can be an option, requiring the bootstrap cluster is a big impediment from a cost and support perspective. This exact issue has come up multiple times in other avenues as well, which maybe it's up to the CAPI community to solve long term. We can take the feedback listed here and act on it within the CAPI project; then we can come back to this issue and as @fabriziopandini mentioned, we can add CAPI to the equation.

It might be worthwhile (and happy to contribute something to this doc) to add a section of future work and alternative considered to list out Cluster API as an overall option, and the gaps we've identified. What do you all think?

Reply:
There is one more aspect to using kOps for testing purposes, which is less complexity. kOps creates the cloud resources starts the K8s cluster and then it gets out of the way. There are no controllers that try to reconcile things in the background. This makes it easy to figure out what is happening with failing tests or broken clusters.

Reply:
@upodroid please consider recording the points above in the "alternatives" section of this KEP. "We don't have one" doesn't seem factual given the discussion above.

Reply:
I'll add a summary of this thread to the alternatives section and join the CAPI meeting this afternoon to say hello and answer some questions.

Reply:
Summarizing the feedback at the Cluster API office hours:

- It'd be great to repurpose the KEP to be focused on improving the current e2e test suites, and not rely on a particular tool to run them (e.g. kube-up or kops).
- Cluster API and the major providers like AWS, GCP, Azure, and vSphere are all currently running conformance test suites. @fabriziopandini will coordinate and open issue with the different providers to start running the other e2e suites like the ones shown by @upodroid in https://testgrid.k8s.io/sig-cluster-lifecycle-kubeup-to-kops
- The Cluster API bootstrap cluster problem can potentially be solved in a number of different ways, which the CAPI community and myself can help with; with the end goal in mind of reducing costs.

@upodroid @CecileRobertMichon @fabriziopandini did I miss anything from the summary?

Reply:
Related note for CAPG maintainers (@dims @cpanato @richardcase): it'd be good to understand what's the delta of features we'd need to run the e2e suites on CAPG. Seems the provider has made lots of progress in the past coming months, and I can carve out some time to help as well.